### PR TITLE
Add bubble check when estimating the depletion voltage

### PR DESCRIPTION
**Pull request by apmypb**

### CODE REVIEW

File: src/Simulation/DepletionVoltage.jl [hunk 1]
```diff
@@ -92,19 +92,25 @@ function estimate_depletion_voltage(sim::Simulation{T},
...
         else
-            maximum(ϕρ[inside] .+ T(U) .* ϕV[inside]) - minimum(ϕρ[inside] .+ T(U) .* ϕV[inside]) < abs(U) ? Umin = U : Umax = U
+            ϕmax - ϕmin < abs(U) ? Umin = U : Umax = U
         end        
     end
-
+    bulk = findall(sim.point_types.data .& bulk_bit .> 0)
+    U_min_max = filter(in(Urng), _find_depletion_voltage_candidates(ϕρ, ϕV, bulk))
+    U2 = isempty(U_min_max) ? U : only(U_min_max)    
+    U = U > 0 ? min(U, U2) : min(U, U2)
```
Comment (fhagemann) [inline src/Simulation/DepletionVoltage.jl:113]: Is this correct? Seems like one of them should be max instead of min (?)

**Comment by fhagemann** [comment]
@zpdg could you please check if this resolves your issue?

**Comment by zpdg** [comment]
> @zpdg could you please check if this resolves your issue?

Of course. This works well after the first "min" was changed to "max" in the `U = U > 0 ? min(U, U2) : min(U, U2)`. And the result is consistent with my code which simply rewrote MJD_siggen's code in julia.
My code: 
```julia
Umin=maximum(broadcast(c -> c.potential, sim.detector.contacts))
Umax=maximum(broadcast(c -> c.potential, sim.detector.contacts))
contact_id=determine_bias_voltage_contact_id(sim.detector)
tolerance=0.1

simDV = deepcopy(sim)
_adapt_weighting_potential_to_electric_potential_grid!(simDV, contact_id)
grid=simDV.electric_potential.grid
ϕT = simDV.electric_potential.data
ϕV = simDV.weighting_potentials[contact_id].data   

bias=simDV.detector.contacts[contact_id].potential
ϕρ=ϕT .-  (bias .* ϕV) # contribution from purities

inside = findall(simDV.point_types .& 4 .> 0)
U::T = 0

while Umax-Umin>tolerance
    U = (Umax + Umin)/2    
    judge=ϕρ[inside] .+ T(U) .* ϕV[inside]
    maximum(judge) - minimum(judge) < abs(U) ? Umax = U : Umin = U
end

dz=[1,-1,0,0]
dr=[0,0,1,-1]

grid_r=grid[1]
grid_z=grid[3]

U2_distance=Umax
for zi in 2:size(ϕT,3)-1
    for ri in 1:size(ϕT,1)-1
        if CartesianIndex(ri,1,zi) in inside
            test_v=-1

            for i in 1:4
                if (ri==1) && (i==3) # don't check the field along the r direction when r==0 mm
                    break
                end

                dV=ϕT[ri+dr[i],1,zi+dz[i]]-ϕT[ri,1,zi]
                dW=ϕV[ri+dr[i],1,zi+dz[i]]-ϕV[ri,1,zi]
                if (dW<0) && (dV<0) && (test_v<dV/dW)
                    test_v=dV/dW
                end
            end

            if (test_v>0) && (U2_distance>test_v)
                U2_distance=test_v
                vminr=ri
                vminz=zi
            end

        end
    end
end
U2=bias-U2_distance
U=max(U,U2)
```

File: test/depletion_test.jl [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    # simulate over and under depletion voltage
+    U₋ = U_est - ΔU
+    U₊ = U_est + ΔU
+    sim.detector = SolidStateDetector(sim.detector, contact_id=1, contact_potential=U₊)
+    calculate_electric_potential!(sim, refinement_limits=0.01)
+    undepleted = is_depleted(sim.point_types)
+    sim.detector = SolidStateDetector(sim.detector, contact_id=1, contact_potential=U₋)
+    calculate_electric_potential!(sim, refinement_limits=0.01)
+    depleted = is_depleted(sim.point_types)
+    @test undepleted && depleted
```
Comment (fhagemann) [inline test/depletion_test.jl:24]: This test will not test the case we are trying to test. We want to add `depletion_handling` here:
```julia
    sim.detector = SolidStateDetector(sim.detector, contact_id=1, contact_potential=U₊)
    calculate_electric_potential!(sim, refinement_limits=0.01, depletion_handling = true)
    undepleted = !is_depleted(sim.point_types)
    sim.detector = SolidStateDetector(sim.detector, contact_id=1, contact_potential=U₋)
    calculate_electric_potential!(sim, refinement_limits=0.01, depletion_handling = true)
    depleted = is_depleted(sim.point_types)
    @test undepleted && depleted
```

File: test/depletion_test.jl [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+using Unitful
+
+T = Float32
+
+@testset "Test depletion estimation" begin
+    sim = Simulation{T}("BEGe_01.yaml")
+    calculate_electric_potential!(sim, refinement_limits=0.01)
+    id = SolidStateDetectors.determine_bias_voltage_contact_id(sim.detector)
+    calculate_weighting_potential!(sim, id, refinement_limits=0.01)
+    SolidStateDetectors._adapt_weighting_potential_to_electric_potential_grid!(sim, id)
```
Comment (fhagemann) [inline test/depletion_test.jl:12]: Is this actually needed in the tests or is it done automatically in `estimate_depletion_voltage` anyhow?

**Review by fhagemann — CHANGES_REQUESTED** [review]

File: test/depletion_test.jl [hunk 4]
```diff
@@ -9,16 +9,21 @@ T = Float32
...
     ΔU = 50
     # simulate over and under depletion voltage
     U₋ = U_est - ΔU
     U₊ = U_est + ΔU
-    sim.detector = SolidStateDetector(sim.detector, contact_id=1, contact_potential=U₊)
+    sim.detector = SolidStateDetector(
+        sim.detector, contact_id=1, contact_potential=U₊, 
+        depletion_handling=true)
     calculate_electric_potential!(sim, refinement_limits=0.01)
     undepleted = is_depleted(sim.point_types)
```
Comment (fhagemann) [inline test/depletion_test.jl:23]: This is still missing an exclamation mark here.
We expect it to be undepleted, so this line should be: !is_depleted

File: test/depletion_test.jl [hunk 5]
```diff
@@ -9,16 +9,21 @@ T = Float32
...
+        sim, id)
     U_est = ustrip(estimate_depletion_voltage(sim)) # around 2600
     ΔU = 50
     # simulate over and under depletion voltage
     U₋ = U_est - ΔU
     U₊ = U_est + ΔU
-    sim.detector = SolidStateDetector(sim.detector, contact_id=1, contact_potential=U₊)
+    sim.detector = SolidStateDetector(
+        sim.detector, contact_id=1, contact_potential=U₊, 
+        depletion_handling=true)
```
Comment (fhagemann) [inline test/depletion_test.jl:21]: The depletion_handling = true needs to be passed to calculate_electric_potential!, not the SolidStateDetector

**Comment by fhagemann** [comment]
Thanks! @apmypb

**Comment by fhagemann** [comment]
I pushed once change that I wanted to incorporate for a long time.
And I renamed the test file to the test naming scheme `test_XXXX.jl`

**Review by fhagemann — APPROVED** [review]

**Comment by fhagemann** [comment]
This will be released soon: https://github.com/JuliaRegistries/General/pull/119509

**Pull request merged**